docs: restructure working PR (#4866)
dfinity/motoko: doc/md/fundamentals/2-types/1-primitive-types.md

let modified = Text.replace("hello world", "world", "Motoko"); // "hello Motoko"
let words = Text.split("apple,banana,cherry", ","); // ["apple", "banana", "cherry"]
```

## Bool

The [`Bool`](https://internetcomputer.org/docs/motoko/base/Bool) type represents boolean values, `true` or `false`, and supports logical operations.

```motoko no-repl
let flag: Bool = Bool.logor(true, false); // true
let opposite = Bool.lognot(flag); // false

let isEqual = Bool.equal(true, false); // false
let comparison = Bool.compare(true, false); // #greater
```

## Float

[`Float`](https://internetcomputer.org/docs/motoko/base/Float) is a 64-bit floating-point type that provides mathematical operations.

```motoko no-repl
import Float "mo:base/Float
let pi = Float.pi;
let radius: Float = 2.5;
let area = Float.pow(radius, 2) * pi; // Area of a circle

let rounded = Float.floor(4.8); // 4.0
let trigValue = Float.sin(Float.pi / 2); // 1.0
```

## Resources

- [`Int`](https://internetcomputer.org/docs/motoko/base/int)
- [`Nat`](https://internetcomputer.org/docs/motoko/base/nat)
- [`Bool`](https://internetcomputer.org/docs/motoko/base/bool)
- [`Char`](https://internetcomputer.org/docs/motoko/base/char)

> - [`Text`](https://internetcomputer.org/docs/motoko/base/text)

Author: @Demali-876

> - [`Text`](https://internetcomputer.org/docs/motoko/base/text)

Reviewer: Resolved in #5062

---
sidebar_position: 1
---

# Primitive types

Motoko provides several primitive types that form the foundation of all computations. These include numeric types, characters and text, booleans, and floating-point numbers.

> 

Reviewer 2: ````suggestion

The primitive types are supported by a large set of familiar built-in operators such `+`, `-` and so on. 

More esoteric functions, not supported by dedicated operators, can be found in the corresponding libraries.

For example, the library function `Int.toText: Int -> Text`, declared in base library `Int`, returns the textual representation of its argument.

``` motoko
import Int "mo:base/Int";
Int.toText(0); // returns "0"
```

````

## Numeric types

Motoko supports both signed integers and unsigned naturals. Signed numbers can represent all numbers, positive and negative, while unsigned integers can only represent 0 and positive numbers. Natural numbers are unsigned integers.

- Signed integers: [`Int`](https://internetcomputer.org/docs/motoko/base/Int), [`Int8`](https://internetcomputer.org/docs/motoko/base/Int8), [`Int16`](https://internetcomputer.org/docs/motoko/base/Int16), [`Int32`](https://internetcomputer.org/docs/motoko/base/Int32), [`Int64`](https://internetcomputer.org/docs/motoko/base/Int64)
- Unsigned naturals: [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat), [`Nat8`](https://internetcomputer.org/docs/motoko/base/Nat8), [`Nat16`](https://internetcomputer.org/docs/motoko/base/Nat16), [`Nat32`](https://internetcomputer.org/docs/motoko/base/Nat32), [`Nat64`](https://internetcomputer.org/docs/motoko/base/Nat64)

> The [`Int`](https://internetcomputer.org/docs/motoko/base/Int) and [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) types prevent overflow and underflow since they dynamically expand or shrink as needed.

Reviewer 2: ```suggestion
The [`Int`](https://internetcomputer.org/docs/motoko/base/Int) and [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) types prevent overflow and underflow since they can represent values of arbitrary size. Of course, subtraction on a `Nat` can still result in underflow if the result would be negative.
```

> While [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) can be implicitly converted to [`Int`](https://internetcomputer.org/docs/motoko/base/Int), the reverse is not always true. [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) is a subtype of [`Int`](https://internetcomputer.org/docs/motoko/base/Int). In other words, a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) value can be used anywhere an [`Int`](https://internetcomputer.org/docs/motoko/base/Int) is expected because [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) is always a valid [`Int`](https://internetcomputer.org/docs/motoko/base/Int). This is why implicit conversions from [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) to [`Int`](https://internetcomputer.org/docs/motoko/base/Int) are allowed.

Reviewer 2: ```suggestion
Since the set of non-negative integers is a subset the integers, in Motoko,  [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) is a subtype of [`Int`](https://internetcomputer.org/docs/motoko/base/Int).
This means that every expression of type 
[`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) can implicitly serve as an  `Int` [`Int`](https://internetcomputer.org/docs/motoko/base/Int) without any need for conversion. In Motoko, it is legal to pass a `Nat` to  a function expecting an `Int`, or to assign a `Nat` to a variable storing an `Int`.
```

```motoko no-repl
let x : Nat = 5;
let y : Int = x; // Allowed
```

> However, an [`Int`](https://internetcomputer.org/docs/motoko/base/Int) cannot be directly assigned to a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) if it holds a negative value, as [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) only supports non-negative numbers.

Reviewer 2: ```suggestion
The converse does not hold.
An [`Int`](https://internetcomputer.org/docs/motoko/base/Int) cannot be directly assigned to a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) since it may be a negative number and the [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) type only contains non-negative numbers.
```

```motoko no-repl
let x : Int = -5;
let y : Nat = x; // Error
```

> To convert a negative [`Int`](https://internetcomputer.org/docs/motoko/base/Int) to a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat), the absolute value must be taken explicitly.

Reviewer 2: ```suggestion
Passing an [`Int`](https://internetcomputer.org/docs/motoko/base/Int) as a [`Nat`](https://internetcomputer.org/docs/motoko/base/Nat) requires a explicit conversion, such as taking the absolute value, or apply some other conversion function.
```

```motoko no-repl
let x : Int = -5;
let y : Nat = Int.abs(x); // Allowed, y = 5
```

Fixed-size numeric types ([`Int8`](https://internetcomputer.org/docs/motoko/base/Int8), [`Nat32`](https://internetcomputer.org/docs/motoko/base/Nat32), etc.) support additional operations, including bitwise shifts.

```motoko no-repl
let x : Nat32 = 0xA; // 10 in hexadecimal

> let y = Nat32.bitshiftLeft(x, 2); // 0x28 (40 in decimal)

Reviewer 2: ```suggestion
let y = x << 2; // 0x28 (40 in decimal)
```

```

## Char and Text

`Char` represents a single Unicode scalar value, while [`Text`](https://internetcomputer.org/docs/motoko/base/Text) represents a sequence of characters.

```motoko no-repl
let letter : Char = 'A';
let codePoint = Char.toNat32(letter); // 65

let word : Text = "Motoko";
let uppercase = Text.toUppercase(word); // "MOTOKO"

let modified = Text.replace("hello world", "world", "Motoko"); // "hello Motoko"
let words = Text.split("apple,banana,cherry", ","); // ["apple", "banana", "cherry"]
```

## Bool

The [`Bool`](https://internetcomputer.org/docs/motoko/base/Bool) type represents boolean values, `true` or `false`, and supports logical operations.

```motoko no-repl

> let flag : Bool = Bool.logor(true, false); // true
> let opposite = Bool.lognot(flag); // false

Reviewer 2: ```suggestion
let flag : Bool = true or false; // true
let opposite = not flag; // false
```

> let isEqual = Bool.equal(true, false); // false
> let comparison = Bool.compare(true, false); // #greater

Reviewer 2: ```suggestion
let isEqual =  true == false ; // false
let comparison = false <= true; // true
```

```

> 

Reviewer 2: ```suggestion
The logical operators `and` and `or` are short-circuiting and will only evaluate their second operand if necessary.
```
